### DISCUSSION TITLE
How to use presentation library to query for content of the whole imodel?

### DISCUSSION THREAD

**Discussion by Snailya** [iTwin/itwinjs-core#3528] [Q&A] [answered]
I want to get a summary table of the imodel.
The Property Formatting Sample only provide a sample on selected element.
I've tried setting the keys with new KeySet() without parameter, but it didn't work.

**Comment by grigasp** [comment, marked as answer]
Can you please explain what is a "summary table of the imodel"?

**Reply by Snailya** [reply]
count of elements

**Reply by ColinKerr** [reply]
ECSql is probably the best way to get a count of elements in an iModel.  Here is an example that gives you the count of elements by class:
```sql
SELECT ECClassId, COUNT(*) as num FROM bis.Element GROUP BY ECClassId ORDER BY COUNT(*) DESC
```

If you want the class name instead of the id use the ec_classname ECSql function: https://www.itwinjs.org/learning/ecsqltutorial/builtinfunctions/#ec_classnameecclassidformat-string--format-id

Is this the type of summary you are looking for?

**Reply by Snailya** [reply]
Indeed, I want to create a merge table like the total volume, the weight of the 304. 316 stainless steel plates. Currently my thought is to store the specification and other material property, e.g. density, into custom item type. Thus, I prefer using presentation rule, which is more flexible that can be stored in json format that could be modified later.

**Reply by ColinKerr** [reply]
Presentation rules is a great way to get the information you need to build a query in this case but it's not an efficient way to retrieve a lot of data from an iModel.  This is the way we build a lot of our UIs internally, we use presentation rules to drive the UI that the user uses to build an ECSql query that actually extracts the data.

Our reporting API is designed to help you solve problems like the one you describe: https://developer.bentley.com/apis/insights/overview/

We also offer Carbon Calculation reporting built on top of the reporting api: https://developer.bentley.com/apis/carbon-calculation/
